<a href="https://colab.research.google.com/github/Tessellate-Imaging/monk_v1/blob/master/study_roadmaps/1_getting_started_roadmap/2_elemental_features_of_monk/7)%20Feature%20-%20Copy%20an%20old%20experiment%20to%20perform%20further%20analysis%20on%20it.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Goals

### 1. Learn where all Copying an experiment feature can be useful
            - When you have a huge dataset, you can divide the data onto large batches, run the first batch for 100 epochs, copy that experiment - take the next batch and run and so on.
            
            - When you have trained a good classifier and do not want to disturb that experiment and yeat try out something new, maybe change hyper-params, etc
            
            - When you train a classifier on huge dataset such as your own custom imagenet and use transfer for a small dataset
            
            - When with time you get new data and want to train on the new dataset, you can copy the previous experiment and train on new data without distrubing the old data
            
            
### Steps
        - Train a classifier using a batch of data
        
        - Copy the experiment and create a new one
        
        - Train the new experiment on extra data


# Table of Contents


## [Install](#0)


## [Train a classifier on one set of data](#1)


## [Copy the experiment](#2)


## [Retrain the new experiment using extra data](#4)

<a id='0'></a>
# Install Monk

## Using pip (Recommended)

  - colab (gpu) 
      - All bakcends: `pip install -U monk-colab`
      

  - kaggle (gpu) 
      - All backends: `pip install -U monk-kaggle`
      

  - cuda 10.2	
      - All backends: `pip install -U monk-cuda102`
      - Gluon bakcned: `pip install -U monk-gluon-cuda102`
	  - Pytorch backend: `pip install -U monk-pytorch-cuda102`
      - Keras backend: `pip install -U monk-keras-cuda102`
      

  - cuda 10.1	
      - All backend: `pip install -U monk-cuda101`
	  - Gluon bakcned: `pip install -U monk-gluon-cuda101`
	  - Pytorch backend: `pip install -U monk-pytorch-cuda101`
	  - Keras backend: `pip install -U monk-keras-cuda101`
      

  - cuda 10.0	
      - All backend: `pip install -U monk-cuda100`
	  - Gluon bakcned: `pip install -U monk-gluon-cuda100`
	  - Pytorch backend: `pip install -U monk-pytorch-cuda100`
	  - Keras backend: `pip install -U monk-keras-cuda100`
      

  - cuda 9.2	
      - All backend: `pip install -U monk-cuda92`
	  - Gluon bakcned: `pip install -U monk-gluon-cuda92`
	  - Pytorch backend: `pip install -U monk-pytorch-cuda92`
	  - Keras backend: `pip install -U monk-keras-cuda92`
      

  - cuda 9.0	
      - All backend: `pip install -U monk-cuda90`
	  - Gluon bakcned: `pip install -U monk-gluon-cuda90`
	  - Pytorch backend: `pip install -U monk-pytorch-cuda90`
	  - Keras backend: `pip install -U monk-keras-cuda90`
      

  - cpu 		
      - All backend: `pip install -U monk-cpu`
	  - Gluon bakcned: `pip install -U monk-gluon-cpu`
	  - Pytorch backend: `pip install -U monk-pytorch-cpu`
	  - Keras backend: `pip install -U monk-keras-cpu`

## Install Monk Manually (Not recommended)
 
### Step 1: Clone the library
 - git clone https://github.com/Tessellate-Imaging/monk_v1.git
 
 
 
 
### Step 2: Install requirements 
 - Linux
     - Cuda 9.0
         - `cd monk_v1/installation/Linux && pip install -r requirements_cu90.txt`
     - Cuda 9.2
         - `cd monk_v1/installation/Linux && pip install -r requirements_cu92.txt`
     - Cuda 10.0
         - `cd monk_v1/installation/Linux && pip install -r requirements_cu100.txt`
     - Cuda 10.1
         - `cd monk_v1/installation/Linux && pip install -r requirements_cu101.txt`
     - Cuda 10.2
         - `cd monk_v1/installation/Linux && pip install -r requirements_cu102.txt`
     - CPU (Non gpu system)
         - `cd monk_v1/installation/Linux && pip install -r requirements_cpu.txt`
 
 
  - Windows
     - Cuda 9.0 (Experimental support)
         - `cd monk_v1/installation/Windows && pip install -r requirements_cu90.txt`
     - Cuda 9.2 (Experimental support)
         - `cd monk_v1/installation/Windows && pip install -r requirements_cu92.txt`
     - Cuda 10.0 (Experimental support)
         - `cd monk_v1/installation/Windows && pip install -r requirements_cu100.txt`
     - Cuda 10.1 (Experimental support)
         - `cd monk_v1/installation/Windows && pip install -r requirements_cu101.txt`
     - Cuda 10.2 (Experimental support)
         - `cd monk_v1/installation/Windows && pip install -r requirements_cu102.txt`
     - CPU (Non gpu system)
         - `cd monk_v1/installation/Windows && pip install -r requirements_cpu.txt`
         
         
  - Mac
     - CPU (Non gpu system)
         - `cd monk_v1/installation/Mac && pip install -r requirements_cpu.txt`
         
         
  - Misc
      - Colab (GPU)
          - `cd monk_v1/installation/Misc && pip install -r requirements_colab.txt`
      - Kaggle (GPU)
          - `cd monk_v1/installation/Misc && pip install -r requirements_kaggle.txt`
 
 
 
### Step 3: Add to system path (Required for every terminal or kernel run)
 - `import sys`
 - `sys.path.append("monk_v1/");`

## Dataset - Malarial cell images
    - Credits: https://www.kaggle.com/iarunava/cell-images-for-detecting-malaria

In [ ]:
! wget --load-cookies /tmp/cookies.txt "https://docs.google.com/uc?export=download&confirm=$(wget --save-cookies /tmp/cookies.txt --keep-session-cookies --no-check-certificate 'https://docs.google.com/uc?export=download&id=1mMEtGIK8UZNCrErXRJR-kutNTaN1zxjC' -O- | sed -rn 's/.*confirm=([0-9A-Za-z_]+).*/\1\n/p')&id=1mMEtGIK8UZNCrErXRJR-kutNTaN1zxjC" -O malaria_cell.zip && rm -rf /tmp/cookies.txt

In [ ]:
! unzip -qq malaria_cell.zipe

In [ ]:
! wget --load-cookies /tmp/cookies.txt "https://docs.google.com/uc?export=download&confirm=$(wget --save-cookies /tmp/cookies.txt --keep-session-cookies --no-check-certificate 'https://docs.google.com/uc?export=download&id=1WHpd7M-E_EiXmdjOr48BfvlUtMRPV6PM' -O- | sed -rn 's/.*confirm=([0-9A-Za-z_]+).*/\1\n/p')&id=1WHpd7M-E_EiXmdjOr48BfvlUtMRPV6PM" -O malaria_cell_val.zip && rm -rf /tmp/cookies.txt

In [ ]:
! unzip -qq malaria_cell_val.zip

# Imports

In [1]:
#Using gluon backend 

# When installed using pip
from monk.gluon_prototype import prototype


# When installed manually (Uncomment the following)
#import os
#import sys
#sys.path.append("monk_v1/");
#sys.path.append("monk_v1/monk/");
#from monk.gluon_prototype import prototype

<a id='1'></a>
# Train a classifier using default settings

### Creating and managing experiments
        - Provide project name
        - Provide experiment name

In [2]:
gtf = prototype(verbose=1);
gtf.Prototype("Malaria-Cell", "first-experiment");

Mxnet Version: 1.5.1

Experiment Details
    Project: Malaria-Cell
    Experiment: first-experiment
    Dir: /home/ubuntu/Desktop/monk_pip_test/monk_v1/study_roadmaps/1_getting_started_roadmap/2_elemental_features_of_monk/workspace/Malaria-Cell/first-experiment/



### This creates files and directories as per the following structure
    
    
    workspace
        |
        |--------Malaria-Cell
                        |
                        |
                        |-----first-experiment
                                    |
                                    |-----experiment-state.json
                                    |
                                    |-----output
                                            |
                                            |------logs (All training logs and graphs saved here)
                                            |
                                            |------models (all trained models saved here)

### Load Dataset

In [3]:
gtf.Default(dataset_path="malaria_cell", 
            model_name="resnet18_v1", 
            num_epochs=5);

#Read the summary generated once you run this cell. 

Dataset Details
    Train path:     malaria_cell
    Val path:       None
    CSV train path: None
    CSV val path:   None
    Label Type:     single

Dataset Params
    Input Size:   224
    Batch Size:   4
    Data Shuffle: True
    Processors:   8
    Train-val split:   0.7

Pre-Composed Train Transforms
[{'RandomHorizontalFlip': {'p': 0.8}}, {'Normalize': {'mean': [0.485, 0.456, 0.406], 'std': [0.229, 0.224, 0.225]}}]

Pre-Composed Val Transforms
[{'RandomHorizontalFlip': {'p': 0.8}}, {'Normalize': {'mean': [0.485, 0.456, 0.406], 'std': [0.229, 0.224, 0.225]}}]

Dataset Numbers
    Num train images: 1411
    Num val images:   605
    Num classes:      2

Model Params
    Model name:           resnet18_v1
    Use Gpu:              True
    Use pretrained:       True
    Freeze base network:  True

Model Details
    Loading pretrained model
    Model Loaded on device
        Model name:                           resnet18_v1
        Num of potentially trainable layers:  41
        Nu

/home/ubuntu/.virtualenvs/monk_pip_test/lib/python3.6/site-packages/monk/system/imports.py:160: UserWarning: ArgumentWarning: clipnorm and clipvalue are active only for keras in current version of Monk
  warnings.warn(msg)
/home/ubuntu/.virtualenvs/monk_pip_test/lib/python3.6/site-packages/monk/system/imports.py:160: UserWarning: ArgumentWarning: momentum_dampening_rate is active only for pytorch in current version of Monk
  warnings.warn(msg)


In [4]:
#Start Training
gtf.Train();

#Read the training summary generated once you run the cell and training is completed

Training Start
    Epoch 1/5
    ----------





    curr_lr - 0.01
    [Epoch 1] Train-acc: 0.725, Train-loss: 0.639 | Val-acc: 0.644628, Val-loss: 0.950, | time: 4.1 sec

    Epoch 2/5
    ----------





    curr_lr - 0.0098
    [Epoch 2] Train-acc: 0.778, Train-loss: 0.571 | Val-acc: 0.809917, Val-loss: 0.448, | time: 4.1 sec

    Epoch 3/5
    ----------





    curr_lr - 0.009604
    [Epoch 3] Train-acc: 0.775, Train-loss: 0.591 | Val-acc: 0.695868, Val-loss: 0.871, | time: 3.9 sec

    Epoch 4/5
    ----------





    curr_lr - 0.009411919999999999
    [Epoch 4] Train-acc: 0.789, Train-loss: 0.547 | Val-acc: 0.742149, Val-loss: 0.649, | time: 3.9 sec

    Epoch 5/5
    ----------




    curr_lr - 0.009223681599999999
    [Epoch 5] Train-acc: 0.795, Train-loss: 0.553 | Val-acc: 0.804959, Val-loss: 0.493, | time: 3.9 sec

    Training completed in: 0m 18s
    Best val Acc:          0.809917

Training End

Training Outputs
    Model Dir:   /home/ubuntu/Desktop/monk_pip_test/monk_v1/study_roadmaps/1_getting_started_roadmap/2_elemental_features_of_monk/workspace/Malaria-Cell/first-experiment/output/models/
    Log Dir:     /home/ubuntu/Desktop/monk_pip_test/monk_v1/study_roadmaps/1_getting_started_roadmap/2_elemental_features_of_monk/workspace/Malaria-Cell/first-experiment/output/logs/
    Final model: final
    Best model:  best_model
    Log 1 - Validation accuracy history log: val_acc_history.npy
    Log 2 - Validation loss history log:     val_loss_history.npy
    Log 3 - Training accuracy history log:   train_acc_history.npy
    Log 4 - Training loss history log:       train_loss_history.npy
    Log 5 - Training curve:                  train_loss_history.npy
   

<Figure size 432x288 with 0 Axes>

## Switch mode from train to eval and Infer

In [5]:
gtf.Switch_Mode(eval_infer=True)

### Load the validation dataset

In [6]:
gtf.Dataset_Params(dataset_path="malaria_cell_val");
gtf.Dataset();

Dataset Details
    Test path:      malaria_cell_val
    CSV test path:  None

Dataset Params
    Input Size:  224
    Processors:   8

Pre-Composed Test Transforms
[{'Normalize': {'mean': [0.485, 0.456, 0.406], 'std': [0.229, 0.224, 0.225]}}]

Dataset Numbers
    Num test images: 1011
    Num classes:      2



### Run validation

In [7]:
accuracy, class_based_accuracy = gtf.Evaluate();

Testing



    Result
        class based accuracies
            0. Parasitized - 40.03944773175542 %
            1. Uninfected - 88.69047619047619 %
        total images:            1011
        num correct predictions: 650
        Average accuracy (%):    64.29277942631059



### Average accuracy post first experiment - 64.29%

<a id='2'></a>
# Copy the experiment

In [8]:
gtf = prototype(verbose=1);
gtf.Prototype("Malaria-Cell", "copied-experiment", 
              copy_from=["Malaria-Cell", "first-experiment"]);

Mxnet Version: 1.5.1

Pre-Composed Train Transforms
[{'RandomHorizontalFlip': {'p': 0.8}}, {'Normalize': {'mean': [0.485, 0.456, 0.406], 'std': [0.229, 0.224, 0.225]}}]

Pre-Composed Val Transforms
[{'RandomHorizontalFlip': {'p': 0.8}}, {'Normalize': {'mean': [0.485, 0.456, 0.406], 'std': [0.229, 0.224, 0.225]}}]

Dataset Numbers
    Num train images: 1411
    Num val images:   605
    Num classes:      2

Model Details
    Loading model - workspace/Malaria-Cell/first-experiment/output/models/final-symbol.json
    Model loaded!

Experiment Details
    Project: Malaria-Cell
    Experiment: copied-experiment
    Dir: /home/ubuntu/Desktop/monk_pip_test/monk_v1/study_roadmaps/1_getting_started_roadmap/2_elemental_features_of_monk/workspace/Malaria-Cell/copied-experiment/



### This creates files and directories as per the following structure
    
    
    workspace
        |
        |--------Malaria-Cell
                        |
                        |
                        |-----first-experiment (Already existant)
                                    |
                                    |-----experiment-state.json
                                    |
                                    |-----output
                                            |
                                            |------logs (All training logs and graphs saved here)
                                            |
                                            |------models (all trained models saved here)
                         |
                         |
                         |-----copied-experiment (Newly created)
                                    |
                                    |-----experiment-state.json
                                    |
                                    |-----output
                                            |
                                            |------logs (All training logs and graphs saved here)
                                            |
                                            |------models (all trained models saved here)

<a id='3'></a>
# Retrain the new experiment using extra data

### Download the dataset

In [ ]:
! wget --load-cookies /tmp/cookies.txt "https://docs.google.com/uc?export=download&confirm=$(wget --save-cookies /tmp/cookies.txt --keep-session-cookies --no-check-certificate 'https://docs.google.com/uc?export=download&id=1BWGkvIWD--m7AOs2of2sDu6TbaktIo-s' -O- | sed -rn 's/.*confirm=([0-9A-Za-z_]+).*/\1\n/p')&id=1BWGkvIWD--m7AOs2of2sDu6TbaktIo-s" -O malaria_cell_extra.zip && rm -rf /tmp/cookies.txt

In [10]:
! unzip -qq malaria_cell_extra.zip

## Load the new dataset
### This part of code will be taken up again in upcoming sections

In [11]:
gtf.update_dataset(dataset_path="malarial_cell_extra");

gtf.Reload();

Pre-Composed Train Transforms
[{'RandomHorizontalFlip': {'p': 0.8}}, {'Normalize': {'mean': [0.485, 0.456, 0.406], 'std': [0.229, 0.224, 0.225]}}]

Pre-Composed Val Transforms
[{'RandomHorizontalFlip': {'p': 0.8}}, {'Normalize': {'mean': [0.485, 0.456, 0.406], 'std': [0.229, 0.224, 0.225]}}]

Dataset Numbers
    Num train images: 621
    Num val images:   69
    Num classes:      2



In [12]:
gtf.Summary()



Experiment Summary

System
    Project Name:    Malaria-Cell
    Project Dir:     workspace/Malaria-Cell/
    Experiment Name: copied-experiment
    Experiment Dir:  workspace/Malaria-Cell/copied-experiment/
    Library:         Mxnet
    Origin:          ['Malaria-Cell', 'first-experiment']

Dataset
    Status:       True
    Dataset Type: train
    Train path:   malarial_cell_extra
    Val path:     None
    Test path:    malaria_cell_val
    CSV Train:    None
    CSV Val:      None
    CSV Test:     False

Dataset Parameters:
    Input Size:   224
    Batch Size:   4
    Shuffle:      True
    Processors:   8
    Num Classes:  2

Dataset Transforms:
    Train transforms: [{'RandomHorizontalFlip': {'p': 0.8}}, {'Normalize': {'mean': [0.485, 0.456, 0.406], 'std': [0.229, 0.224, 0.225]}}]
    Val transforms:   [{'RandomHorizontalFlip': {'p': 0.8}}, {'Normalize': {'mean': [0.485, 0.456, 0.406], 'std': [0.229, 0.224, 0.225]}}]
    Test transforms:  [{'Normalize': {'mean': [0.485, 0.45

### In the summary, it points to its copied Origin:          ['Malaria-Cell', 'first-experiment']

In [13]:
#Start Training
gtf.Train();

#Read the training summary generated once you run the cell and training is completed

Training Start
    Epoch 1/5
    ----------





    curr_lr - 0.01
    [Epoch 1] Train-acc: 0.894, Train-loss: 0.371 | Val-acc: 0.985507, Val-loss: 0.068, | time: 2.5 sec

    Epoch 2/5
    ----------





    curr_lr - 0.0098
    [Epoch 2] Train-acc: 0.913, Train-loss: 0.268 | Val-acc: 0.985507, Val-loss: 0.121, | time: 2.4 sec

    Epoch 3/5
    ----------





    curr_lr - 0.009604
    [Epoch 3] Train-acc: 0.939, Train-loss: 0.183 | Val-acc: 0.898551, Val-loss: 0.294, | time: 2.4 sec

    Epoch 4/5
    ----------





    curr_lr - 0.009411919999999999
    [Epoch 4] Train-acc: 0.948, Train-loss: 0.121 | Val-acc: 1.000000, Val-loss: 0.028, | time: 2.6 sec

    Epoch 5/5
    ----------





    curr_lr - 0.009223681599999999
    [Epoch 5] Train-acc: 0.973, Train-loss: 0.054 | Val-acc: 1.000000, Val-loss: 0.008, | time: 2.6 sec

    Training completed in: 0m 10s
    Best val Acc:          1.000000

Training End

Training Outputs
    Model Dir:   /home/ubuntu/Desktop/monk_pip_test/monk_v1/study_roadmaps/1_getting_started_roadmap/2_elemental_features_of_monk/workspace/Malaria-Cell/copied-experiment/output/models/
    Log Dir:     /home/ubuntu/Desktop/monk_pip_test/monk_v1/study_roadmaps/1_getting_started_roadmap/2_elemental_features_of_monk/workspace/Malaria-Cell/copied-experiment/output/logs/
    Final model: final
    Best model:  best_model
    Log 1 - Validation accuracy history log: val_acc_history.npy
    Log 2 - Validation loss history log:     val_loss_history.npy
    Log 3 - Training accuracy history log:   train_acc_history.npy
    Log 4 - Training loss history log:       train_loss_history.npy
    Log 5 - Training curve:                  train_loss_history.npy


<Figure size 432x288 with 0 Axes>

In [14]:
gtf.Summary()



Experiment Summary

System
    Project Name:    Malaria-Cell
    Project Dir:     workspace/Malaria-Cell/
    Experiment Name: copied-experiment
    Experiment Dir:  workspace/Malaria-Cell/copied-experiment/
    Library:         Mxnet
    Origin:          ['Malaria-Cell', 'first-experiment']

Dataset
    Status:       True
    Dataset Type: train
    Train path:   malarial_cell_extra
    Val path:     None
    Test path:    malaria_cell_val
    CSV Train:    None
    CSV Val:      None
    CSV Test:     False

Dataset Parameters:
    Input Size:   224
    Batch Size:   4
    Shuffle:      True
    Processors:   8
    Num Classes:  2

Dataset Transforms:
    Train transforms: [{'RandomHorizontalFlip': {'p': 0.8}}, {'Normalize': {'mean': [0.485, 0.456, 0.406], 'std': [0.229, 0.224, 0.225]}}]
    Val transforms:   [{'RandomHorizontalFlip': {'p': 0.8}}, {'Normalize': {'mean': [0.485, 0.456, 0.406], 'std': [0.229, 0.224, 0.225]}}]
    Test transforms:  [{'Normalize': {'mean': [0.485, 0.45

## Switch mode from train to eval and validate

In [15]:
gtf.Switch_Mode(eval_infer=True)

### Load the validation dataset

In [16]:
# This part of code will be taken up again in upcoming sections
gtf.Dataset_Params(dataset_path="malaria_cell_val");
gtf.reset_transforms(test=True);
gtf.apply_normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225], test=True);
gtf.Dataset();

Dataset Details
    Test path:      malaria_cell_val
    CSV test path:  None

Dataset Params
    Input Size:  224
    Processors:   8

Pre-Composed Test Transforms
[{'Normalize': {'mean': [0.485, 0.456, 0.406], 'std': [0.229, 0.224, 0.225]}}]

Dataset Numbers
    Num test images: 1011
    Num classes:      2



### Run validation

In [17]:
accuracy, class_based_accuracy = gtf.Evaluate();

Testing



    Result
        class based accuracies
            0. Parasitized - 99.40828402366864 %
            1. Uninfected - 83.53174603174604 %
        total images:            1011
        num correct predictions: 925
        Average accuracy (%):    91.49357072205737



### Average accuracy post copied experiment - 91.493%

# Goals Completed

### 1. Learn where all Copying an experiment feature can be useful
            
### Steps learnt
        - Train a classifier using a batch of data
        
        - Copy the experiment and create a new one
        
        - Train the new experiment on extra data
